src/main.rs

```diff
@@ -252,149 +254,164 @@ fn main() -> Result<()> {
-                    }
-                    _ => (),
+    // Emit documentation if requested.
+    if let Some(dir) = matches.value_of("docdir") {
+        info!("Generating documentation in `{}`", dir);
+        let doc = doc::Doc::new(&syntax_trees);
+        let mut html = doc::Renderer::new(Path::new(dir));
+        html.render(&doc)?;
+        return Ok(());
+    }
+
+    for pf in &syntax_trees {
+        for node in &pf.ast {
+            trace!("{:?}", node);
+            match node {
+                // Module declarations.
+                RefNode::ModuleDeclarationAnsi(x) => {
+                    // unwrap_node! gets the nearest ModuleIdentifier from x
+                    let id = unwrap_node!(x, SimpleIdentifier).unwrap();
+                    pickle.register_declaration(&pf.ast, id);
                 }
+                RefNode::ModuleDeclarationNonansi(x) => {
+                    let id = unwrap_node!(x, SimpleIdentifier).unwrap();
+                    pickle.register_declaration(&pf.ast, id);
+                }
+                // Interface Declaration.
+                RefNode::InterfaceDeclaration(x) => {
+                    let id = unwrap_node!(x, SimpleIdentifier).unwrap();
+                    pickle.register_declaration(&pf.ast, id);
+                }
+                // Package declarations.
+                RefNode::PackageDeclaration(x) => {
+                    let id = unwrap_node!(x, SimpleIdentifier).unwrap();
+                    pickle.register_declaration(&pf.ast, id);
+                }
+                _ => (),
             }
-            // Iterate again and check for usage
-            for node in &syntax_tree {
-                match node {
-                    // Instantiations, end-labels.
-                    RefNode::ModuleIdentifier(x) => {
-                        let id = unwrap_node!(x, SimpleIdentifier).unwrap();
-                        pickle.register_usage(&syntax_tree, id);
-                    }
-                    // Interface identifier.
-                    RefNode::InterfaceIdentifier(x) => {
-                        let id = unwrap_node!(x, SimpleIdentifier).unwrap();
-                        pickle.register_usage(&syntax_tree, id);
-                    }
-                    // Package Qualifier (i.e., explicit package constants).
-                    RefNode::ClassScope(x) => {
-                        let id = unwrap_node!(x, SimpleIdentifier).unwrap();
-                        pickle.register_usage(&syntax_tree, id);
-                    }
-                    // Package Import.
-                    RefNode::PackageIdentifier(x) => {
-                        let id = unwrap_node!(x, SimpleIdentifier).unwrap();
-                        pickle.register_usage(&syntax_tree, id);
-                    }
-                    _ => (),
+        }
+    }
+    for pf in &syntax_trees {
+        // For each file, start with a clean replacement table.
+        pickle.replace_table = vec![];
```

Comment:
You can save 10ns here by just clearing the table, which keeps the allocated backing memory intact :stuck_out_tongue: 
```suggestion
        pickle.replace_table.clear();
```

Reply:
Very nice input, thanks! Better safe some instructions here 😄